File: shims/qpid-jms/src/main/java/org/apache/qpid/interop_test/shim/AmqpReceiver.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.qpid.interop_test.shim;

import java.math.BigDecimal;
import java.util.UUID;
import java.util.Vector;
import javax.jms.BytesMessage;
import javax.jms.Connection;
import javax.jms.ConnectionFactory;
import javax.jms.ExceptionListener;
import javax.jms.JMSException;
import javax.jms.MapMessage;
import javax.jms.Message;
import javax.jms.MessageConsumer;
import javax.jms.ObjectMessage;
import javax.jms.Queue;
import javax.jms.Session;
import javax.jms.StreamMessage;
import javax.jms.TextMessage;
import javax.naming.Context;
import javax.naming.InitialContext;
import org.apache.qpid.jms.JmsConnectionFactory;

public class AmqpReceiver {
    private static final String USER = "guest";
    private static final String PASSWORD = "guest";
    private static final int TIMEOUT = 1000;
    private static final String[] SUPPORTED_AMQP_TYPES = {"null",
													      "boolean",
													      "ubyte",
													      "ushort",
													      "uint",
													      "ulong",
													      "byte",
													      "short",
													      "int",
													      "long",
													      "float",
													      "double",
													      "decimal32",
													      "decimal64",
													      "decimal128",
													      "char",
													      "timestamp",
													      "uuid",
													      "binary",
													      "string",
													      "symbol",
													      "list",
													      "map",
													      "array"};

    public static void main(String[] args) throws Exception {
    	if (args.length < 4) {
    		System.out.println("AmqpReceiver: Insufficient number of arguments");
    		System.out.println("AmqpReceiver: Expected arguments: broker_address, queue_name, amqp_type, num_test_values");
    		System.exit(1);
    	}
    	String brokerAddress = "amqp://" + args[0];
    	String queueName = args[1];
    	String amqpType = args[2];
    	int numTestValues = Integer.parseInt(args[3]);
    	Connection connection = null;

        try {
        	ConnectionFactory factory = (ConnectionFactory)new JmsConnectionFactory(brokerAddress);

            connection = factory.createConnection(USER, PASSWORD);
            connection.setExceptionListener(new MyExceptionListener());
            connection.start();

            Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
            
            Queue queue = session.createQueue(queueName);

            MessageConsumer messageConsumer = session.createConsumer(queue);
            
            Vector<String> outList = new Vector<String>();
            outList.add(amqpType);
            if (isSupportedAmqpType(amqpType)) {
                int actualCount = 0;
                Message message = null;
                for (int i = 1; i <= numTestValues; i++, actualCount++) {
        			message = messageConsumer.receive(TIMEOUT);
        			if (message == null)
        				break;
            		switch (amqpType) {
            		case "null":
            			long bodyLength = ((BytesMessage)message).getBodyLength();
            			if (bodyLength == 0L) {
            				outList.add("None");
            			} else {
            				throw new Exception("AmqpReceiver: JMS BytesMessage size error: Expected 0 bytes, read " + bodyLength);
            			}
            			break;
            		case "boolean":
            			String bs = String.valueOf(((BytesMessage)message).readBoolean());
            			outList.add(Character.toUpperCase(bs.charAt(0)) + bs.substring(1));
            			break;
            		case "ubyte":
            			byte byteValue = ((BytesMessage)message).readByte();
            			short ubyteValue = (short)(byteValue & 0xff);
            			outList.add(String.valueOf(ubyteValue));
            			break;
            		case "ushort":
            		{
            			byte[] byteArray = new byte[2];
            			int numBytes = ((BytesMessage)message).readBytes(byteArray);
            			if (numBytes != 2) {
            				// TODO: numBytes == -1 means no more bytes in stream - add error message for this case?
            				throw new Exception("AmqpReceiver: JMS BytesMessage size error: Exptected 2 bytes, read " + numBytes);
            			}
            			int ushortValue = 0;
            			for (int j=0; j<byteArray.length; j++) {
            				ushortValue = (ushortValue << 8) + (byteArray[j] & 0xff);
            			}
            			outList.add(String.valueOf(ushortValue));
            			break;
            		}
            		case "uint":
            		{
            			byte[] byteArray = new byte[4];
            			int numBytes = ((BytesMessage)message).readBytes(byteArray);
            			if (numBytes != 4) {
            				// TODO: numBytes == -1 means no more bytes in stream - add error message for this case?
            				throw new Exception("AmqpReceiver: JMS BytesMessage size error: Exptected 4 bytes, read " + numBytes);
            			}
            			long uintValue = 0;
            			for (int j=0; j<byteArray.length; j++) {
            				uintValue = (uintValue << 8) + (byteArray[j] & 0xff);
            			}
            			outList.add(String.valueOf(uintValue));
            			break;
            		}
            		case "ulong":
            		case "timestamp":
            		{
            			// TODO: Tidy this ugliness up - perhaps use of vector<byte>?
            			byte[] byteArray = new byte[8];
            			int numBytes = ((BytesMessage)message).readBytes(byteArray);
            			if (numBytes != 8) {
            				// TODO: numBytes == -1 means no more bytes in stream - add error message for this case?
            				throw new Exception("AmqpReceiver: JMS BytesMessage size error: Exptected 8 bytes, read " + numBytes);
            			}
            			// TODO: shortcut in use here - this byte array should go through a Java type that can represent this as a number - such as BigInteger.
            			outList.add(String.format("0x%02x%02x%02x%02x%02x%02x%02x%02x", byteArray[0], byteArray[1],
            					byteArray[2], byteArray[3], byteArray[4], byteArray[5], byteArray[6], byteArray[7]));
            			break;
            		}
            		case "byte":
            			outList.add(String.valueOf(((BytesMessage)message).readByte()));
            			break;
            		case "short":
            			outList.add(String.valueOf(((BytesMessage)message).readShort()));
            			break;
            		case "int":
            			outList.add(String.valueOf(((BytesMessage)message).readInt()));
            			break;
            		case "long":
            			outList.add(String.valueOf(((BytesMessage)message).readLong()));
            			break;
            		case "float":
            			float f = ((BytesMessage)message).readFloat();
            			int i0 = Float.floatToRawIntBits(f);
            			outList.add(String.format("0x%8s", Integer.toHexString(i0)).replace(' ', '0'));
            			break;
            		case "double":
            			double d = ((BytesMessage)message).readDouble();
            			long l = Double.doubleToRawLongBits(d);
            			outList.add(String.format("0x%16s", Long.toHexString(l)).replace(' ', '0'));
            			break;
            		case "decimal32":
            			BigDecimal bd32 = (BigDecimal)((ObjectMessage)message).getObject();
            			outList.add(bd32.toString());
            			break;
            		case "decimal64":
            			BigDecimal bd64 = (BigDecimal)((ObjectMessage)message).getObject();
            			outList.add(bd64.toString());
            			break;
            		case "decimal128":
            			BigDecimal bd128 = (BigDecimal)((ObjectMessage)message).getObject();
            			outList.add(bd128.toString());
            			break;
            		case "char":
            			outList.add(String.format("%c", ((BytesMessage)message).readChar()));
            			break;
            		case "uuid":
            			UUID uuid = (UUID)((ObjectMessage)message).getObject();
            			outList.add(uuid.toString());
            			break;
            		case "binary":
            			BytesMessage bm = (BytesMessage)message;
            			int msgLen = (int)bm.getBodyLength();
            			byte[] ba = new byte[msgLen];
            			if (bm.readBytes(ba) == msgLen) {
            				outList.add(new String(ba));
            			} else {
            				// TODO: Raise exception or error here: size mismatch
            			}
            			break;
            		case "string":
            			outList.add(((TextMessage)message).getText());
            			break;
            		case "symbol":
            			outList.add(((BytesMessage)message).readUTF());
            			break;
            		case "list":
            			break;
            		case "map":
            			break;
            		case "array":
            			break;
            		default:
            			// Internal error, should never happen if SUPPORTED_AMQP_TYPES matches this case stmt
            			connection.close();
            			throw new Exception("AmqpReceiver: Internal error: unsupported AMQP type \"" + amqpType + "\"");
            		}
                }
            } else {
            	System.out.println("ERROR: AmqpReceiver: AMQP type \"" + amqpType + "\" is not supported");
            	connection.close();
            	System.exit(1);
            }

            connection.close();

            // No exception, print results
            for (int i=0; i<outList.size(); i++) {
            	System.out.println(outList.get(i));
            }
        } catch (Exception exp) {
        	if (connection != null)
        		connection.close();
            System.out.println("Caught exception, exiting.");
            exp.printStackTrace(System.out);
            System.exit(1);
        }
    }
    
    protected static boolean isSupportedAmqpType(String amqpType) {
    	for (String supportedAmqpType: SUPPORTED_AMQP_TYPES) {
    		if (amqpType.equals(supportedAmqpType))
    			return true;
    	}
    	return false;
    }

    private static class MyExceptionListener implements ExceptionListener {
        @Override
        public void onException(JMSException exception) {
            System.out.println("Connection ExceptionListener fired, exiting.");
            exception.printStackTrace(System.out);
            System.exit(1);
        }
    }
}

File: shims/qpid-proton-cpp/src/qpidit/shim/AmqpReceiver.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include "qpidit/shim/AmqpReceiver.hpp"

#include <iostream>
#include <json/json.h>
#include <proton/types.hpp>
#include "proton/container.hpp"
#include "proton/event.hpp"
#include "qpidit/QpidItErrors.hpp"

namespace qpidit
{
    namespace shim
    {

        AmqpReceiver::AmqpReceiver(const std::string& brokerUrl,
                                   const std::string& amqpType,
                                   uint32_t expected) :
                        _brokerUrl(brokerUrl),
                        _amqpType(amqpType),
                        _expected(expected),
                        _received(0UL),
                        _receivedValueList(Json::arrayValue)
        {}

        AmqpReceiver::~AmqpReceiver() {}

        Json::Value& AmqpReceiver::getReceivedValueList() {
            return _receivedValueList;
        }

        void AmqpReceiver::on_container_start(proton::event &e, proton::container &c) {
            _receiver = c.open_receiver(_brokerUrl);
        }

        void AmqpReceiver::on_message(proton::event &e, proton::delivery &d, proton::message &m) {
            if (m.id().get<uint64_t>() < _received) return; // ignore duplicate
            if (_received < _expected) {
                if (_amqpType.compare("null") == 0) {
                    checkMessageType(m, proton::NULL_TYPE);
                    _receivedValueList.append("None");
                } else if (_amqpType.compare("boolean") == 0) {
                    checkMessageType(m, proton::BOOLEAN);
                    _receivedValueList.append(m.body().get<bool>() ? "True": "False");
                } else if (_amqpType.compare("ubyte") == 0) {
                    checkMessageType(m, proton::UBYTE);
                    _receivedValueList.append(toHexStr<uint8_t>(m.body().get<uint8_t>()));
                } else if (_amqpType.compare("ushort") == 0) {
                    checkMessageType(m, proton::USHORT);
                    _receivedValueList.append(toHexStr<uint16_t>(m.body().get<uint16_t>()));
                } else if (_amqpType.compare("uint") == 0) {
                    checkMessageType(m, proton::UINT);
                    _receivedValueList.append(toHexStr<uint32_t>(m.body().get<uint32_t>()));
                } else if (_amqpType.compare("ulong") == 0) {
                    checkMessageType(m, proton::ULONG);
                    _receivedValueList.append(toHexStr<uint64_t>(m.body().get<uint64_t>()));
                } else if (_amqpType.compare("byte") == 0) {
                    checkMessageType(m, proton::BYTE);
                    _receivedValueList.append(toHexStr<int8_t>(m.body().get<int8_t>()));
                } else if (_amqpType.compare("short") == 0) {
                    checkMessageType(m, proton::SHORT);
                    _receivedValueList.append(toHexStr<int16_t>(m.body().get<int16_t>()));
                } else if (_amqpType.compare("int") == 0) {
                    checkMessageType(m, proton::INT);
                    _receivedValueList.append(toHexStr<int32_t>(m.body().get<int32_t>()));
                } else if (_amqpType.compare("long") == 0) {
                    checkMessageType(m, proton::LONG);
                    _receivedValueList.append(toHexStr<int64_t>(m.body().get<int64_t>()));
                } else if (_amqpType.compare("float") == 0) {
                    checkMessageType(m, proton::FLOAT);
                    float f = m.body().get<float>();
                    _receivedValueList.append(toHexStr<uint32_t>(*((uint32_t*)&f), true));
                } else if (_amqpType.compare("double") == 0) {
                    checkMessageType(m, proton::DOUBLE);
                    double d = m.body().get<double>();
                    _receivedValueList.append(toHexStr<uint64_t>(*((uint64_t*)&d), true));
                } else if (_amqpType.compare("decimal32") == 0) {
                    checkMessageType(m, proton::DECIMAL32);
                    _receivedValueList.append(byteArrayToHexStr(m.body().get<proton::decimal32>()));
                } else if (_amqpType.compare("decimal64") == 0) {
                    checkMessageType(m, proton::DECIMAL64);
                    _receivedValueList.append(byteArrayToHexStr(m.body().get<proton::decimal64>()));
                } else if (_amqpType.compare("decimal128") == 0) {
                    checkMessageType(m, proton::DECIMAL128);
                    _receivedValueList.append(byteArrayToHexStr(m.body().get<proton::decimal128>()));
                } else if (_amqpType.compare("char") == 0) {
                    checkMessageType(m, proton::CHAR);
                    wchar_t c = m.body().get<wchar_t>();
                    std::stringstream oss;
                    if (c < 0x7f && std::iswprint(c)) {
                        oss << (char)c;
                    } else {
                        oss << "0x" << std::hex << c;
                    }
                    _receivedValueList.append(oss.str());
                } else if (_amqpType.compare("timestamp") == 0) {
                    checkMessageType(m, proton::TIMESTAMP);
                    std::ostringstream oss;
                    oss << "0x" << std::hex << m.body().get<proton::timestamp>().milliseconds();
                    _receivedValueList.append(oss.str());
                } else if (_amqpType.compare("uuid") == 0) {
                    checkMessageType(m, proton::UUID);
                    std::ostringstream oss;
                    oss << m.body().get<proton::uuid>();
                    _receivedValueList.append(oss.str());
                } else if (_amqpType.compare("binary") == 0) {
                    checkMessageType(m, proton::BINARY);
                    _receivedValueList.append(std::string(m.body().get<proton::binary>()));
                } else if (_amqpType.compare("string") == 0) {
                    checkMessageType(m, proton::STRING);
                    _receivedValueList.append(m.body().get<std::string>());
                } else if (_amqpType.compare("symbol") == 0) {
                    checkMessageType(m, proton::SYMBOL);
                    _receivedValueList.append(m.body().get<proton::symbol>());
                } else if (_amqpType.compare("list") == 0) {
                    checkMessageType(m, proton::LIST);
                    Json::Value jsonList(Json::arrayValue);
                    _receivedValueList.append(getSequence(jsonList, m.body()));
                } else if (_amqpType.compare("map") == 0) {
                    checkMessageType(m, proton::MAP);
                    Json::Value jsonMap(Json::objectValue);
                    _receivedValueList.append(getMap(jsonMap, m.body()));
                } else if (_amqpType.compare("array") == 0) {
                    throw qpidit::UnsupportedAmqpTypeError(_amqpType);
                } else {
                    throw qpidit::UnknownAmqpTypeError(_amqpType);
                }
            }
            _received++;
            if (_received >= _expected) {
                e.receiver().close();
                e.connection().close();
            }
        }

        void AmqpReceiver::on_connection_error(proton::event &e, proton::connection &c) {
            std::cerr << "AmqpReceiver:on_connection_error() event=" << e.name() << std::endl;
        }

        void AmqpReceiver::on_sender_error(proton::event &e, proton::sender& l) {
            std::cerr << "AmqpReceiver:on_sender_error() event=" << e.name() << std::endl;
        }

        void AmqpReceiver::on_transport_error(proton::event &e, proton::transport &t) {
            std::cerr << "AmqpReceiver:on_transport_error() event=" << e.name() << std::endl;
        }

        void AmqpReceiver::on_unhandled_error(proton::event &e, const proton::condition &c) {
            std::cerr << "AmqpReceiver:on_unhandled_error() event=" << e.name() << " condition=" << c.name() << std::endl;
        }

        // protected

        //static
        void AmqpReceiver::checkMessageType(const proton::message& msg, proton::type_id amqpType) {
            if (msg.body().type() != amqpType) {
                throw qpidit::IncorrectMessageBodyTypeError(amqpType, msg.body().type());
            }
        }

        //static
        Json::Value& AmqpReceiver::getMap(Json::Value& jsonMap, const proton::value& val) {
            std::map<proton::value, proton::value> msgMap;
            val.get(msgMap);
            for (std::map<proton::value, proton::value>::const_iterator i = msgMap.begin(); i != msgMap.end(); ++i) {
                switch (i->second.type()) {
                case proton::LIST:
                {
                    Json::Value jsonSubList(Json::arrayValue);
                    jsonMap[i->first.get<std::string>()] = getSequence(jsonSubList, i->second);
                    break;
                }
                case proton::MAP:
                {
                    Json::Value jsonSubMap(Json::objectValue);
                    jsonMap[i->first.get<std::string>()] = getMap(jsonSubMap, i->second);
                    break;
                }
                case proton::ARRAY:
                    break;
                case proton::STRING:
                    jsonMap[i->first.get<std::string>()] = Json::Value(i->second.get<std::string>());
                    break;
                default:
                    throw qpidit::IncorrectValueTypeError(i->second);
                }
            }
            return jsonMap;
        }

        //static
        Json::Value& AmqpReceiver::getSequence(Json::Value& jsonList, const proton::value& val) {
            std::vector<proton::value> msgList;
            val.get(msgList);
            for (std::vector<proton::value>::const_iterator i=msgList.begin(); i!=msgList.end(); ++i) {
                switch ((*i).type()) {
                case proton::LIST:
                {
                    Json::Value jsonSubList(Json::arrayValue);
                    jsonList.append(getSequence(jsonSubList, *i));
                    break;
                }
                case proton::MAP:
                {
                    Json::Value jsonSubMap(Json::objectValue);
                    jsonList.append(getMap(jsonSubMap, *i));
                    break;
                }
                case proton::ARRAY:
                    break;
                case proton::STRING:
                    jsonList.append(Json::Value((*i).get<std::string>()));
                    break;
                default:
                    throw qpidit::IncorrectValueTypeError(*i);
                }
            }
            return jsonList;
        }

        //static
        std::string AmqpReceiver::stringToHexStr(const std::string& str) {
            std::ostringstream oss;
            oss << "0x" << std::hex;
            for (std::string::const_iterator i=str.begin(); i!=str.end(); ++i) {
                oss << std::setw(2) << std::setfill('0') << ((int)*i & 0xff);
            }
            return oss.str();
        }

    } /* namespace shim */
} /* namespace qpidit */


/*
 * --- main ---
 * Args: 1: Broker address (ip-addr:port)
 *       2: Queue name
 *       3: AMQP type
 *       4: Expected number of test values to receive
 */

int main(int argc, char** argv) {
    // TODO: improve arg management a little...
    if (argc != 5) {
        throw qpidit::ArgumentError("Incorrect number of arguments");
    }

    std::ostringstream oss;
    oss << argv[1] << "/" << argv[2];

    try {
        qpidit::shim::AmqpReceiver receiver(oss.str(), argv[3], std::strtoul(argv[4], NULL, 0));
        proton::container(receiver).run();

        std::cout << argv[3] << std::endl;
        std::cout << receiver.getReceivedValueList();
    } catch (const std::exception& e) {
        std::cerr << "AmqpReceiver error: " << e.what() << std::endl;
        exit(-1);
    }
    exit(0);
}
